### PR TITLE
Add heuristic for choosing the "preferred" provider

### PR DESCRIPTION
**Pull request by iainmcgin**
The behavior of the heuristic is specified in #7.

### CODE REVIEW

File: api/java/org/openyolo/api/CredentialClient.java [hunk 1]
```diff
@@ -222,6 +250,63 @@ public Credential getCredentialFromActivityResult(@Nullable Intent resultData) {
         return responders;
     }
 
+    @Nullable
+    private ComponentName getPreferredProvider(@NonNull List<ComponentName> providers) {
```
Comment (dxslly) [inline api/java/org/openyolo/api/CredentialClient.java]: Is it your intention to skip the picker when there are untrusted providers and one preferred provider?

Comment (iainmcgin) [inline api/java/org/openyolo/api/CredentialClient.java]: Very good point. Amended the heuristic such that if there are any unknown providers installed, there is no preferred provider.

File: api/java/org/openyolo/api/KnownProviders.java [hunk 2]
```diff
@@ -124,6 +121,27 @@ public static KnownProviders getApplicationBoundInstance(Context context) {
         return providers;
     }
 
+    /**
```
Comment (dxslly) [inline api/java/org/openyolo/api/KnownProviders.java]: We could avoid adding this test only methods by using a shadow.  I think it is preferable in this case due to the potential risk of these methods' usage.

For example:
public class ShadowKnownProviders {
   private static KnowProviders mImplemenation;
   public static void provideImpl() {...}
   public static KnownProviders getApplicationBoundInstance(Context context) {
       return mImplemenation;
   }
}

Comment (iainmcgin) [inline api/java/org/openyolo/api/KnownProviders.java]: I'm not clear on when shadows can and cannot be used. I'll do some reading.

**Pull request merged**